### PR TITLE
validate components in blueprint

### LINKED ISSUE #503: Validate blueprint component in using `validate_or`
We are using `?` operator to get each component of `Blueprint { schema, definitions, server, upstream }`.

This exits with the first error that's produced. Instead we should use `validate_or` to validate all of them independently and then combine the result to produce the Blueprint.

**Tech Requirements**
- Add unit tests to make sure all errors a captured an combined together.

### PR DESCRIPTION
**Pull request by wrath-of-god**
/claim #503

**Issue Reference(s):**  
Fixes #503 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh` to address and fix linting issues.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have performed a self-review of my own code.

### CODE REVIEW

**Comment by wrath-of-god** [comment]
@tusharmath @amitksingh1490 , codecov failing test passes, review and rerun codecov

File: src/blueprint/from_config.rs [hunk 1]
```diff
@@ -29,12 +29,15 @@ type Valid<A> = ValidDefault<A, String>;
...
+  let definitions = to_definitions(config, output_types, input_types);
+  let server: Valid<Server> = Server::try_from(config.server.clone());
   let upstream = config.upstream.clone();
-  valid_base_url(upstream.base_url.as_ref())?;
-  let blueprint = Blueprint { schema, definitions, server, upstream };
+  let server = valid_base_url(upstream.base_url.as_ref())
+    .validate_or(schema.clone())
+    .validate_or(definitions.clone())
+    .validate_or(server)?;
+  let blueprint = Blueprint { schema: schema?, definitions: definitions?, server, upstream };
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:40]: Come to think of it, I think we should use `validate_both` and then a `map` operator to combine and create a blueprint.

```rust
let (schema, definitions, server, upstream) = schema
  .validate_both(definitions)
  .validate_both(server)
  .validate_both(upstream)
  .map(|tuple| tuple.flatten())?
  
```

Checkout
https://crates.io/crates/tuples

Comment (wrath-of-god) [reply src/blueprint/from_config.rs:40]: why need flatten its not necessary computation
let (((_, schema), definitions), server) can destructure

File: src/blueprint/from_config.rs [hunk 2]
```diff
@@ -29,11 +29,14 @@ type Valid<A> = ValidDefault<A, String>;
...
-  let definitions = to_definitions(config, output_types, input_types)?;
-  let server: Server = Server::try_from(config.server.clone())?;
+  let schema = to_schema(config);
+  let definitions = to_definitions(config, output_types, input_types);
+  let server: Valid<Server> = Server::try_from(config.server.clone());
   let upstream = config.upstream.clone();
-  valid_base_url(upstream.base_url.as_ref())?;
+  let (((_, schema), definitions), server) = valid_base_url(upstream.base_url.as_ref())
+    .validate_both(schema.clone())
+    .validate_both(definitions.clone())
```
Comment (meskill) [inline src/blueprint/from_config.rs:38]: do we actually need the `.clone()` here? seems like we are overriding those variables and it's fine to just pass ownership of the original value

Comment (wrath-of-god) [reply src/blueprint/from_config.rs]: No not needed, removing

File: src/blueprint/from_config.rs [hunk 3]
```diff
@@ -29,11 +29,14 @@ type Valid<A> = ValidDefault<A, String>;
...
-  let server: Server = Server::try_from(config.server.clone())?;
+  let schema = to_schema(config);
+  let definitions = to_definitions(config, output_types, input_types);
+  let server: Valid<Server> = Server::try_from(config.server.clone());
   let upstream = config.upstream.clone();
-  valid_base_url(upstream.base_url.as_ref())?;
+  let (((_, schema), definitions), server) = valid_base_url(upstream.base_url.as_ref())
+    .validate_both(schema)
+    .validate_both(definitions)
+    .validate_both(server)?;
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:39]: Why not use `flatten` in https://crates.io/crates/tuples

```suggestion
  let (_, schema, definitions, server) = valid_base_url(upstream.base_url.as_ref())
    .validate_both(schema)
    .validate_both(definitions)
    .validate_both(server).map(|a| a.flatten())?;
```

**Pull request closed without merging**